File: src/dotsboxes/game.py
from .board import Board


class TurnResult:

    def __init__(self, current_player: int, last_player: int, last_filled: list = []):
        self.current_player = current_player
        self.last_player = last_player
        self.last_filled = last_filled

    def __eq__(self, other: object) -> bool:
        if isinstance(other, TurnResult):
            return (self.current_player == other.current_player and
                    self.last_player == other.last_player and
                    self.last_filled == other.last_filled)

        return False

    def __str__(self) -> str:
        return (f'TurnResult(current_player={self.current_player}, '
                f'last_player={self.last_player}, '
                f'last_filled={self.last_filled})'
                )


class Game:

    def __init__(self, board: Board, player_count: int, current_player=1):
        if player_count < 0 or player_count > 4:
            raise ValueError(
                'player_count must be greater than zero and no more than four.')

        self.board = board
        self.player_count = player_count
        self.current_player = current_player

    def mark_line(self, row: int, column: int):
        last_player = self.current_player
        filled = self.board.mark_line(row, column, self.current_player)

        if filled is not None and not filled:
            self.current_player = self.current_player % self.player_count + 1

        return TurnResult(self.current_player, last_player, filled)

    def outcome(self):
        if self.board.has_open_boxes():
            return []

        scores = [(player, self.board.fill_count(player))
                  for player in range(1, self.player_count + 1)]
        scores.sort(key=lambda s: s[1], reverse=True)

        return scores

File: core/src/main/java/com/hutsondev/dotsboxes/core/Game.java
package com.hutsondev.dotsboxes.core;

import java.util.List;
import java.util.Optional;
import lombok.Getter;
import lombok.NonNull;

public class Game {

  private final Board board;

  @Getter
  private Player currentPlayer;

  public Game(@NonNull Board board, Player currentPlayer) {
    this.board = board;
    this.currentPlayer = currentPlayer;
  }

  public Game(@NonNull Board board) {
    this(board, Player.ONE);
  }

  public Game(int rowCount, int columnCount) {
    this(new Board(rowCount, columnCount));
  }

  public BoardView board() {
    return board;
  }

  public TurnResult markLine(int row, int column) {
    Player lastPlayer = currentPlayer;

    Optional<List<Integer>> filled = board.markLine(row, column, currentPlayer);

    if (filled.isPresent() && filled.get().isEmpty()) {
      currentPlayer = currentPlayer.other();
    }

    return new TurnResult(currentPlayer, lastPlayer, filled);
  }

  public Optional<Outcome> getOutcome() {
    if (board.hasOpenBoxes()) {
      return Optional.empty();
    }

    return Optional.of(new Outcome(
        board.filledBoxCount(Player.ONE),
        board.filledBoxCount(Player.TWO)
    ));
  }
}
